### PR TITLE
[CALCITE-2791]Add the JSON_TYPE function

### PR DESCRIPTION
**Pull request by XuQianJin-Stars**
https://issues.apache.org/jira/browse/CALCITE-2791

### CODE REVIEW

**Comment by XuQianJin-Stars** [comment]
This error is broken by other tests

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java [hunk 1]
```diff
@@ -2662,6 +2662,37 @@ public static void jsonArrayAggAdd(List list, Object element,
...
+        result = "DATE";
+      } else if (o instanceof Map) {
+        result = "OBJECT";
+      } else if (o instanceof Collection) {
+        result = "ARRAY";
+      } else if (o == null) {
+        result = "NULL";
+      }
+    } catch (Exception ex) {
+      throw new CalciteException(ex.getMessage(), ex.getCause());
```
Comment (zhztheplayer) [inline core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java]: The exception should be thrown using the class RESOURCE.java, you can find some usage examples just in this file.

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java [hunk 2]
```diff
@@ -2662,6 +2662,37 @@ public static void jsonArrayAggAdd(List list, Object element,
     }
   }
 
+  public static String jsonType(Object input) {
+    String result = "";
```
Comment (zhztheplayer) [inline core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java]: Would it be better to make variable `result` final and assign values in if branches? So it need not to be assigned twice.

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java [hunk 3]
```diff
@@ -2662,6 +2662,37 @@ public static void jsonArrayAggAdd(List list, Object element,
...
+      if (o instanceof Integer) {
+        result = "INTEGER";
+      } else if (o instanceof String) {
+        result = "STRING";
+      } else if (o instanceof Float) {
+        result = "FLOAT";
+      } else if (o instanceof Double) {
+        result = "DOUBLE";
+      } else if (o instanceof Long) {
+        result = "LONG";
```
Comment (zhztheplayer) [inline core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java]: Is it really needed to distinguish INTEGER and LONG (So do FLOAT and DOUBLE)? I see MySQL's JSON_TYPE[1] does not provide LONG and DOUBLE output.

[1] https://dev.mysql.com/doc/refman/8.0/en/json-attribute-functions.html#function_json-type

File: core/src/main/java/org/apache/calcite/sql/fun/SqlJsonTypeFunction.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+/**
+ * The <code>JSON_TYPE</code> function.
+ */
+public class SqlJsonTypeFunction extends SqlFunction {
+  public SqlJsonTypeFunction() {
+    super("JSON_TYPE",
+            SqlKind.OTHER_FUNCTION,
+            ReturnTypes.explicit(SqlTypeName.VARCHAR, 20),
+            null,
+            OperandTypes.VARIADIC,
```
Comment (zhztheplayer) [inline core/src/main/java/org/apache/calcite/sql/fun/SqlJsonTypeFunction.java]: OperandTypes.VARIADIC indicates a parameter list with varying length. Bu it looks like  JSON_TYPE is designed for processing a single input value.
Is OperandTypes.ANY enough for that?

File: core/src/main/java/org/apache/calcite/sql/fun/SqlJsonTypeFunction.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+            SqlFunctionCategory.SYSTEM);
+  }
+
+  @Override public SqlOperandCountRange getOperandCountRange() {
+    return SqlOperandCountRanges.from(1);
```
Comment (zhztheplayer) [inline core/src/main/java/org/apache/calcite/sql/fun/SqlJsonTypeFunction.java]: The same. Maybe something like `SqlOperandCountRanges.of(1)` is enough.

File: core/src/main/java/org/apache/calcite/sql/fun/SqlJsonTypeFunction.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+                                             SqlOperandTypeChecker argType, SqlCall call) {
+    assert call.operandCount() == 1;
+  }
+
+  @Override public SqlCall createCall(SqlLiteral functionQualifier,
+                                      SqlParserPos pos, SqlNode... operands) {
+    return super.createCall(functionQualifier, pos, operands);
+  }
+
+  @Override public void unparse(SqlWriter writer, SqlCall call, int leftPrec,
```
Comment (zhztheplayer) [inline core/src/main/java/org/apache/calcite/sql/fun/SqlJsonTypeFunction.java]: It seems that the following does the same thing with `super.unparse(...)`.

File: core/src/main/java/org/apache/calcite/sql/fun/SqlJsonTypeFunction.java [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+                                int rightPrec) {
+    assert call.operandCount() == 1;
+    final SqlWriter.Frame frame = writer.startFunCall(getName());
+    call.operand(0).unparse(writer, 0, 0);
+    writer.endFunCall(frame);
+  }
+
+  private <E extends Enum<E>> E getEnumValue(SqlNode operand) {
```
Comment (zhztheplayer) [inline core/src/main/java/org/apache/calcite/sql/fun/SqlJsonTypeFunction.java]: This method seems to be redundant.

File: core/src/test/java/org/apache/calcite/sql/test/SqlOperatorBaseTest.java [hunk 8]
```diff
@@ -4473,6 +4473,32 @@ private void checkNullOperand(SqlTester tester, String op) {
         "{\"foo\":{\"foo\":\"bar\"}}", "VARCHAR(2000) NOT NULL");
   }
 
+  @Test public void testJsonType() {
+    tester.setFor(SqlStdOperatorTable.JSON_TYPE);
+    tester.checkString("json_type('1')",
```
Comment (zhztheplayer) [inline core/src/test/java/org/apache/calcite/sql/test/SqlOperatorBaseTest.java]: I guess the right case should be 
```java
tester.checkString("json_type('\\"1\\"')", "INTEGER", "VARCHAR(20) NOT NULL");
```

**Comment by XuQianJin-Stars** [comment]
hi @zhztheplayer  Thank you very much. I benefit a lot from your review. I will start to revise it right away.
best
qianjin

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java (hunk 3 above)
Comment (XuQianJin-Stars) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java]: A Long can be generated if:
json_type(timestamp '2016-02-24 12:42:25')
The double type contains the float scene so it is converted to a double.
and now Binary types and OPAQUE types: Jackson handle would give you an error.

**Comment by XuQianJin-Stars** [comment]
hi @zhztheplayer  I have revised this PR. I hope you can spare your precious time to review it
best
qianjin

File: core/src/main/java/org/apache/calcite/runtime/CalciteResource.java [hunk 9]
```diff
@@ -849,6 +849,9 @@
 
   @BaseMessage("While executing SQL [{0}] on JDBC sub-schema")
   ExInst<RuntimeException> exceptionWhilePerformingQueryOnJdbcSubSchema(String sql);
+
+  @BaseMessage("Unkonw json type in JSON_TYPE function, and the object is: ''{0}''")
```
Comment (zhztheplayer) [inline core/src/main/java/org/apache/calcite/runtime/CalciteResource.java]: Here are some typos:
Unkonw -> Unknown
json -> JSON

File: core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java [hunk 10]
```diff
@@ -10769,6 +10769,12 @@ private void checkCustomColumnResolving(String table) {
         "(?s).*Expected a character type*");
   }
 
+  @Test public void testJsonType() {
+    check("select json_type(ename) from emp");
+    checkExp("json_type('{\"foo\":\"bar\"}')");
+    checkExpType("json_type('{\"foo\":\"bar\"}')", "VARCHAR(20) NOT NULL");
```
Comment (zhztheplayer) [inline core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java]: I prefer adding a case for something like `json_type(1)`, which should fail the validation.

File: site/_docs/reference.md [hunk 11]
```diff
@@ -567,6 +567,7 @@ JSON,
 **JSON_OBJECT**,
 **JSON_OBJECTAGG**,
 **JSON_QUERY**,
+**JSON_TYPE**,
```
Comment (zhztheplayer) [inline site/_docs/reference.md]: Is `JSON_TYPE` currently a reserved keyword of MySQL?

And you may want to document usage of `JSON_TYPE` also. Something like: http://calcite.apache.org/docs/reference.html#query-functions

Comment (zhztheplayer) [reply site/_docs/reference.md]: I recall that something like \*\*JSON_TYPE\*\* indicates a reserved keyword.

If you don't want make a keyword reserved, maybe you should add it to https://github.com/apache/calcite/blob/master/core/src/main/codegen/templates/Parser.jj#L6690

Comment (XuQianJin-Stars) [reply site/_docs/reference.md]: I've added non-reserved keywords

Comment (zhztheplayer) [reply site/_docs/reference.md]: Could you please add non-reserved keyword `JSON_TYPE` to here or nearby:

https://github.com/apache/calcite/blob/d9fe7f7ad7c2006780e459ce90e3f442a6246051/core/src/main/codegen/config.fmpp#L161

File: site/_docs/reference.md [hunk 12]
```diff
@@ -1992,6 +1993,24 @@ Note:
 | value IS JSON ARRAY                               | Whether *value* is a json array, *value* is in character string type
 | value IS NOT JSON ARRAY                           | Whether *value* is not a json array, *value* is in character string type
 
+#### Other Operators
```
Comment (zhztheplayer) [inline site/_docs/reference.md]: Could you please make it clear that the operator `JSON_TYPE` is from MySQL?
My suggestion is to change "Other Operators" to "MySQL Specific Operators"

File: site/_docs/reference.md (hunk 11 above)
Comment (XuQianJin-Stars) [reply site/_docs/reference.md]: hi @zhztheplayer My attempts to use non-reserved keywords resulted in many test conflicts and too many changes. I'll keep the keywords
best
qianjin

File: site/_docs/reference.md [hunk 13]
```diff
@@ -1992,6 +1993,24 @@ Note:
...
+
+Example Data:
+
+```JSON
+{"a": [10, true]}
+```
+
+Comparison:
+
+|Operator                                    |{"a": [10, true]}|$.a          |$.a[0]          |$.a[1]
```
Comment (zhztheplayer) [inline site/_docs/reference.md]: Could a single `$` better than `{"a": [10, true]}` here?

**Comment by XuQianJin-Stars** [comment]
hi @zhztheplayer  I have revised this PR. I hope you can spare your precious time to review it
best
qianjin

File: site/_docs/reference.md [hunk 14]
```diff
@@ -2002,14 +2002,14 @@ Note:
 Example Data:
 
 ```JSON
-{"a": [10, true]}
+{"a": [10, true],"b": "[10, true]"}
```
Comment (zhztheplayer) [inline site/_docs/reference.md]: @XuQianJin-Stars, Sorry for confusing you in the [JIRA discussion](https://issues.apache.org/jira/browse/CALCITE-2791?focusedCommentId=16773621&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-16773621). Let's remove line 2004 - line 2006 and just put the SQL text here, what do you think?

File: site/_docs/reference.md [hunk 15]
```diff
@@ -2002,14 +2002,14 @@ Note:
...
 ```
 
 Comparison:
```
Comment (zhztheplayer) [inline site/_docs/reference.md]: i think here we can remove this line and following lines, and put the result set directly.

File: site/_docs/reference.md (hunk 14 above)
Comment (XuQianJin-Stars) [reply site/_docs/reference.md]: hi @zhztheplayer  Well, that's a good idea.
best
qianjin

**Comment by XuQianJin-Stars** [comment]
hi @zhztheplayer  I have addressed all comments.
best
qianjin

File: site/_docs/reference.md [hunk 16]
```diff
@@ -1992,6 +1993,29 @@ Note:
...
+
+Example SQL:
+
+```SQl
+SELECT JSON_TYPE(v) AS c1
+,JSON_TYPE(JSON_VALUE(v, 'lax $.b' ERROR ON ERROR)) AS c2
+,JSON_TYPE(JSON_VALUE(v, 'strict $.a[0]' ERROR ON ERROR)) AS c3
+,JSON_TYPE(JSON_VALUE(v, 'strict $.a[1]' ERROR ON ERROR)) AS c4
+FROM (VALUES ('{"a": [10, true],"b": "[10, true]"}')) AS t(v)
+limit 10;
```
Comment (zhztheplayer) [inline site/_docs/reference.md]: Let's keep upper case convention for keyword `LIMIT`.

Comment (XuQianJin-Stars) [reply site/_docs/reference.md]: well,let me change it soon.

File: site/_docs/reference.md [hunk 17]
```diff
@@ -1992,6 +1993,29 @@ Note:
...
 
+#### MySQL Specific Operators
+
+| Operator syntax                                   | Description
+|:------------------------------------------------- |:-----------
+| JSON_TYPE(value) | Returns a string indicating the type of a JSON **value**. This can be an object, an array, or a scalar type
+
+Example SQL:
+
+```SQl
```
Comment (zhztheplayer) [inline site/_docs/reference.md]: Could you please fix typo: `SQl` -> `SQL`

Comment (XuQianJin-Stars) [reply site/_docs/reference.md]: Thank you very much for helping me review again and again.I have already change it.

**Comment by zhztheplayer** [comment]
@XuQianJin-Stars, following is details about how you can make `JSON_TYPE` non-reserved:
1. Add `JSON_TYPE` to list `nonReservedKeywords` in /root/Documents/SourceCode/calcite/babel/src/main/codegen/config.fmpp;
2. Add `JSON_TYPE` to list `nonReservedKeywords` in /root/Documents/SourceCode/calcite/core/src/main/codegen/config.fmpp;
3. Add `JSON_TYPE` to list `nonReservedKeywords` in /root/Documents/SourceCode/calcite/core/src/test/codegen/config.fmpp;
4. Add `JSON_TYPE` to list `nonReservedKeywords` in /root/Documents/SourceCode/calcite/server/src/main/codegen/config.fmpp;
5. Revert your changes about keywords in SqlAdvisorTest and SqlParserTest;
6. Change `**JSON_TYPE**` to `JSON_TYPE` from reference.md.

**Comment by XuQianJin-Stars** [comment]
hi @zhztheplayer  Thank you very much. I have added `JSON_TYPE`into the non-reserved keyword according to your help.
best
qianjin

**Comment by zhztheplayer** [comment]
@XuQianJin-Stars Could you please solve the conflicts and squash your commits into a single commit?

**Comment by XuQianJin-Stars** [comment]
hi @zhztheplayer I'm on the train. I'll deal with this at night. Thank you for your understanding. 
best
qianjin

**Pull request closed without merging**